src/MinimiseProgramTransformer.cpp

```diff
@@ -39,12 +40,134 @@
+    std::set<std::string> constants{};
+    std::vector<NormalisedClauseElementRepr> clauseElements;
+
+    /**
+     * Parse an atom with a preset name qualifier into the element list.
+     */
+    void addClauseAtom(std::string qualifier, const AstAtom* atom);
+
+    /**
+     * Parse a body literal into the element list.
+     */
+    void addClauseBodyLiteral(const AstLiteral* lit);
+
+    /**
+     * Return a normalised string repr of an argument.
+     */
+    std::string normaliseArgument(const AstArgument* arg);
+};
+
+void MinimiseProgramTransformer::NormalisedClauseRepr::addClauseAtom(
+        std::string qualifier, const AstAtom* atom) {
+    AstQualifiedName name(atom->getQualifiedName());
+    name.prepend(qualifier);
+
+    std::vector<std::string> vars;
+    for (const auto* arg : atom->getArguments()) {
+        vars.push_back(normaliseArgument(arg));
+    }
+    clauseElements.push_back({.name = name, .params = vars});
+}
+
+void MinimiseProgramTransformer::NormalisedClauseRepr::addClauseBodyLiteral(const AstLiteral* lit) {
+    if (const auto* atom = dynamic_cast<const AstAtom*>(lit)) {
+        addClauseAtom("@min:atom", atom);
+    } else if (const auto* neg = dynamic_cast<const AstNegation*>(lit)) {
+        addClauseAtom("@min:neg", neg->getAtom());
+    } else if (const auto* bc = dynamic_cast<const AstBinaryConstraint*>(lit)) {
+        AstQualifiedName name(toBinaryConstraintSymbol(bc->getOperator()));
+        name.prepend("@min:operator");
+        std::vector<std::string> vars;
+        vars.push_back(normaliseArgument(bc->getLHS()));
+        vars.push_back(normaliseArgument(bc->getRHS()));
+        clauseElements.push_back({.name = name, .params = vars});
+    } else {
+        fullyNormalised = false;
+        AstQualifiedName name(toString(*lit));
+        name.prepend("@min:unhandled:lit");
+        clauseElements.push_back({.name = name, .params = std::vector<std::string>()});
+    }
+}
+
+std::string MinimiseProgramTransformer::NormalisedClauseRepr::normaliseArgument(const AstArgument* arg) {
+    if (auto* stringCst = dynamic_cast<const AstStringConstant*>(arg)) {
+        std::stringstream name;
+        name << "@min:cst:str" << *stringCst;
+        constants.insert(name.str());
+        return name.str();
+    } else if (auto* numericCst = dynamic_cast<const AstNumericConstant*>(arg)) {
+        std::stringstream name;
+        name << "@min:cst:num:" << *numericCst;
+        constants.insert(name.str());
+        return name.str();
+    } else if (dynamic_cast<const AstNilConstant*>(arg) != nullptr) {
+        constants.insert("@min:cst:nil");
+        return "@min:cst:nil";
+    } else if (auto* var = dynamic_cast<const AstVariable*>(arg)) {
+        auto name = var->getName();
+        variables.insert(name);
+        return name;
+    } else if (dynamic_cast<const AstUnnamedVariable*>(arg)) {
+        static size_t countUnnamed = 0;
+        std::string name = "@min:unnamed:" + countUnnamed++;
```

Comment:
This is still incrementing the char*. You could try something like
```suggestion
        std::string name = "@min:unnamed:" + std::to_string(countUnnamed++);
```